contracts/token/LinearVestingWithPenalty.sol

```diff
@@ -415,37 +280,122 @@ contract LinearVestingWithPenalty is Initializable, ERC20Upgradeable, AccessCont
+                totalReleasable += releasable;
+            }
+        }
 
-        VestingSchedule[] storage fromSchedules = vestingSchedules[from];
-        uint256 schedulesCount = fromSchedules.length;
-        if (schedulesCount == 0) revert NoVestingSchedule();
+        require(totalReleasable > 0, "No tokens available for release");
+
+        _burn(msg.sender, totalReleasable);
+        sophtoken.safeTransfer(msg.sender, totalReleasable);
+        emit TokensReleased(msg.sender, totalReleasable, totalReleasable, 0);
+    }
 
-        VestingSchedule[] storage toSchedules = vestingSchedules[to];
-        for (uint256 i = 0; i < schedulesCount; i++) {
-            toSchedules.push(fromSchedules[i]);
+    /**
+    * @dev Returns the list of unclaimed vesting schedule indexes for the beneficiary within a specified range and their respective releasable amounts.
+    * @param beneficiary The address of the beneficiary to check unclaimed schedules for.
+    * @param start The starting index of the range (inclusive).
+    * @param end The ending index of the range (exclusive).
+    * @return indexes An array of indexes for schedules with unclaimed tokens within the specified range.
+    * @return amounts An array of unclaimed amounts corresponding to each index within the specified range.
+    */
+    function getUnclaimedSchedulesInRange(
+        address beneficiary,
+        uint256 start,
+        uint256 end
+    ) external view returns (uint256[] memory indexes, uint256[] memory amounts) {
+        uint256 scheduleCount = vestingSchedules[beneficiary].length;
+        require(start < end && end <= scheduleCount, "Invalid range");
+
+        // Create temporary arrays with a fixed maximum size (end - start)
+        uint256[] memory tempIndexes = new uint256[](end - start);
+        uint256[] memory tempAmounts = new uint256[](end - start);
+        uint256 unclaimedCount = 0;
+
+        // Single loop to collect unclaimed schedules within the range
+        for (uint256 i = start; i < end; i++) {
+            VestingSchedule storage schedule = vestingSchedules[beneficiary][i];
+            uint256 releasable = _releasableAmount(schedule);
+
+            if (releasable > 0) {
+                tempIndexes[unclaimedCount] = i;
+                tempAmounts[unclaimedCount] = releasable;
+                unclaimedCount++;
+            }
         }
 
-        delete vestingSchedules[from];
+        // Create result arrays with exact size of unclaimed schedules
+        indexes = new uint256[](unclaimedCount);
+        amounts = new uint256[](unclaimedCount);
+
+        // Copy collected data to result arrays
+        for (uint256 j = 0; j < unclaimedCount; j++) {
+            indexes[j] = tempIndexes[j];
+            amounts[j] = tempAmounts[j];
+        }
+
+        return (indexes, amounts);
+    }
+
+
+    /**
+    * @dev Claims the entire releasable amount for specific vesting schedule indexes with an applied penalty.
```

Comment:
Penalty shouldn't apply to what's releasable as per my understanding. Releasable has already vested and it should be freely claimable, penalty should apply to non-vested amount (total - released - releasable).

`penaltyAmount = ((schedule.totalAmount - schedule.released - releasable) * penalty percentage) / 100
`

Reply:
good point

Reply:
This is a reference of how claiming arbitrary non vested amount would work:

```
    // nonVestedFractions works as a percentage of what's non vested that will be claimed WITH PENALTY
    function claimSpecificSchedulesWithPenalty(uint256[] calldata scheduleIndexes, uint256[] calldata nonVestedFractions) external {
        if (vestingStartDate > block.timestamp) revert VestingHasNotStartedYet();

        uint256 totalReleasable = 0;
        uint256 totalPenaltyAmount = 0;
        
        // Check for array consistency
        if (scheduleIndexes.length != nonVestedFractions.length) revert MismatchedArrayLengths();

        for (uint256 i = 0; i < scheduleIndexes.length; i++) {
            uint256 index = scheduleIndexes[i];
            if (index >= vestingSchedules[msg.sender].length) revert InvalidScheduleIndex();

            VestingSchedule storage schedule = vestingSchedules[msg.sender][index];
            uint256 releasable = _releasableAmount(schedule);
            
            if (releasable > 0) {      
                uint256 nonVestedFraction = nonVestedFractions[i];

                // Check that the non releasable portion is a percentage, that is, it's less than or equal to 100
                // (Optional: by default 100 or bigger could mean claim all the nonVested tokens with penalty)
                if (nonVestedFraction > 100) {
                    revert InvalidnonVestedProportion();

                // Uses does not want a non releasable portion (0%)
                } else if (nonVestedFraction == 0) {
                    schedule.released += releasable;
                    totalReleasable += releasable;
                    
                } else {
                    // User want's a given percentage of the non relesable tokens. We calculate them.
                    // schedule.totalAmount - schedule.release - releasable is the nonVested amount
                    uint256 nonVestedFractionAmount = ((schedule.totalAmount - schedule.release - releasable) * nonVestedFraction) / 100;
                    // Apply a penalty over those tokens.
                    uint256 penaltyAmount = (nonVestedFractionAmount * penaltyPercentage) / 100;
                    // User claims whatever is releasable plus what is nonVested minus the penalty
                    uint256 netAmount = releasable + nonVestedFractionAmount - penaltyAmount;

                    totalReleasable += netAmount;
                    totalPenaltyAmount += penaltyAmount;
                    schedule.released += netAmount;
                }
            }
        }

        require(totalReleasable > 0, "No tokens available for release");

        if (totalPenaltyAmount > 0) {
            sophtoken.safeTransfer(penaltyRecipient, totalPenaltyAmount);
            emit PenaltyPaid(msg.sender, totalPenaltyAmount);
        }

        sophtoken.safeTransfer(msg.sender, totalReleasable);
        emit TokensReleased(msg.sender, totalReleasable + totalPenaltyAmount, totalReleasable, totalPenaltyAmount);
        _burn(msg.sender, totalReleasable + totalPenaltyAmount);
    }
```